{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "c6aea1a4",
   "metadata": {},
   "source": [
    "## Context dependent models with CpG effect\n",
    "\n",
    "CpG effect is deamination of 5-methy-cytosine at both positions, i.e. the C and the G (which is a C on the opposite strand). Thus the change is CpG->TpG OR CpG->CpA.\n",
    "\n",
    "To complete the definition of the context dependent model, we also include the general nucleotide parameters.\n",
    "\n",
    "### Relationship to the independent (nucleotide) process\n",
    "\n",
    "If you omit the CpG parameters below and provide the dinucleotide state frequencies as the Kronecker product of nucleotide frequencies, then the log-likelihood of that model will be identical to the GN model with the same nucleotide frequencies.\n",
    "\n",
    "> **Note**\n",
    "> The branch length for the dinucleotide model is the expected number of substitutions per aligned *dinucleotide*. If memory serves correct, this will be exactly 1/2 that from the nucleotide model. So all genetic distances (the branch lengths in ENS) you are using need to be transformed into the expected number of substitutions per aligned *nucleotide*."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "b83a30c7",
   "metadata": {},
   "outputs": [],
   "source": [
    "from itertools import permutations\n",
    "import cogent3 as c3\n",
    "from cogent3.evolve.predicate import MotifChange\n",
    "from cogent3.evolve.ns_substitution_model import NonReversibleDinucleotide, NonReversibleCodon\n",
    "\n",
    "def make_gn_preds():\n",
    "    # making the model parameters (predictates) for\n",
    "    # the General Nucleotide Markov model\n",
    "    return [\n",
    "        MotifChange(f, t, forward_only=True)\n",
    "        for f, t in permutations(\"ACTG\", 2)\n",
    "        if f != \"T\" or t != \"G\"\n",
    "    ]\n",
    "\n",
    "def make_nr_cpg_preds_strand_asymetric():\n",
    "    # strand specific CpG deamination rates, so separate\n",
    "    # parameters for plus strand (CG->TG) and minus\n",
    "    # strand (CG->CA)\n",
    "    return [\n",
    "        MotifChange(\"CG\", \"TG\", forward_only=True),\n",
    "        MotifChange(\"CG\", \"CA\", forward_only=True),\n",
    "    ]\n",
    "\n",
    "def make_nr_cpg_preds_strand_symetric():\n",
    "    # same CpG deamination rate on both strands\n",
    "    # so one parameter\n",
    "    return [\n",
    "        # | is the binary or operator that combines the predicates\n",
    "        # so we have the union of the two changes as a single parameter\n",
    "        MotifChange(\"CG\", \"TG\", forward_only=True) |\n",
    "        MotifChange(\"CG\", \"CA\", forward_only=True)\n",
    "    ]\n",
    "\n",
    "def _make_model(cls, **kwargs):\n",
    "    return cls(**kwargs)\n",
    "\n",
    "def GDN_CpG_ss(**kwargs):\n",
    "    \"\"\"return a dinucleotide model with strand symmetric CpG deamination\"\"\"\n",
    "    ssym_preds = make_gn_preds() + make_nr_cpg_preds_strand_symetric()\n",
    "    kwargs=dict(predicates=ssym_preds, optimise_motif_probs=True, name=\"GDN_CpG_ss\")\n",
    "    return _make_model(NonReversibleDinucleotide, **kwargs)\n",
    "\n",
    "def GDN_CpG(**kwargs):\n",
    "    \"\"\"return a dinucleotide model with strand symmetric CpG deamination\"\"\"\n",
    "    asym_preds = make_gn_preds() + make_nr_cpg_preds_strand_asymetric()\n",
    "    kwargs=dict(predicates=asym_preds, optimise_motif_probs=True, name=\"GDN_CpG\")\n",
    "    return _make_model(NonReversibleDinucleotide, **kwargs)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "bcba2e89",
   "metadata": {},
   "outputs": [],
   "source": [
    "aln = c3.get_dataset(\"primate-brca1\")\n",
    "tree = c3.get_dataset(\"primate-tree\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "86bb2064",
   "metadata": {},
   "source": [
    "we drop gapped columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "ccc25510",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<style>\n",
       ".c3align table {margin: 10px 0;}\n",
       ".c3align td { border: none !important; text-align: left !important; }\n",
       ".c3align tr:not(.num_row) td span {margin: 0 2px;}\n",
       ".c3align tr:nth-child(even) {background: #f7f7f7;}\n",
       ".c3align .num_row {background-color:rgba(161, 195, 209, 0.5) !important; border-top: solid 1px black; }\n",
       ".c3align .label { font-size: 12pt ; text-align: right !important; color: black !important; padding: 0 4px; display: table-cell !important; font-weight: normal !important; }\n",
       ".c3align .T_dna{font-family: \"Lucida Console\",monospace !important; font-size: 12pt !important; color: blue; }\n",
       ".c3align .C_dna{font-family: \"Lucida Console\",monospace !important; font-size: 12pt !important; color: black; }\n",
       ".c3align .A_dna{font-family: \"Lucida Console\",monospace !important; font-size: 12pt !important; color: #FF0102; }\n",
       ".c3align .G_dna{font-family: \"Lucida Console\",monospace !important; font-size: 12pt !important; color: green; }\n",
       ".c3align .terminal_ambig_dna{font-family: \"Lucida Console\",monospace !important; font-size: 12pt !important; color: gray; }\n",
       ".c3align .ambig_dna{font-family: \"Lucida Console\",monospace !important; font-size: 12pt !important; color: gray; }\n",
       "</style>\n",
       "<div class=\"c3align\">\n",
       "<table>\n",
       "<tr class=\"num_row\"><td></td><td><b>0</b></td></tr>\n",
       "<tr><td class=\"label\">Chimpanzee</td><td><span class=\"T_dna\">T</span><span class=\"G_dna\">G</span><span class=\"T_dna\">T</span><span class=\"G_dna\">G</span><span class=\"G_dna\">G</span><span class=\"C_dna\">C</span><span class=\"A_dna\">A</span><span class=\"C_dna\">C</span><span class=\"A_dna\">A</span><span class=\"A_dna\">A</span><span class=\"A_dna\">A</span><span class=\"T_dna\">T</span><span class=\"A_dna\">A</span><span class=\"C_dna\">C</span><span class=\"T_dna\">T</span><span class=\"C_dna\">C</span><span class=\"A_dna\">A</span><span class=\"T_dna\">T</span><span class=\"G_dna\">G</span><span class=\"C_dna\">C</span><span class=\"C_dna\">C</span><span class=\"A_dna\">A</span><span class=\"G_dna\">G</span><span class=\"C_dna\">C</span><span class=\"T_dna\">T</span><span class=\"C_dna\">C</span><span class=\"A_dna\">A</span><span class=\"T_dna\">T</span><span class=\"T_dna\">T</span><span class=\"A_dna\">A</span><span class=\"C_dna\">C</span><span class=\"A_dna\">A</span><span class=\"G_dna\">G</span><span class=\"C_dna\">C</span><span class=\"A_dna\">A</span><span class=\"T_dna\">T</span><span class=\"G_dna\">G</span><span class=\"A_dna\">A</span><span class=\"G_dna\">G</span><span class=\"A_dna\">A</span><span class=\"A_dna\">A</span><span class=\"C_dna\">C</span><span class=\"A_dna\">A</span><span class=\"G_dna\">G</span><span class=\"C_dna\">C</span><span class=\"A_dna\">A</span><span class=\"G_dna\">G</span><span class=\"T_dna\">T</span><span class=\"T_dna\">T</span><span class=\"T_dna\">T</span><span class=\"A_dna\">A</span><span class=\"T_dna\">T</span><span class=\"T_dna\">T</span><span class=\"A_dna\">A</span><span class=\"C_dna\">C</span><span class=\"T_dna\">T</span><span class=\"C_dna\">C</span><span class=\"A_dna\">A</span><span class=\"C_dna\">C</span><span class=\"T_dna\">T</span></td></tr>\n",
       "<tr><td class=\"label\">Human</td><td><span class=\"T_dna\">.</span><span class=\"G_dna\">.</span><span class=\"T_dna\">.</span><span class=\"G_dna\">.</span><span class=\"G_dna\">.</span><span class=\"C_dna\">.</span><span class=\"A_dna\">.</span><span class=\"C_dna\">.</span><span class=\"A_dna\">.</span><span class=\"A_dna\">.</span><span class=\"A_dna\">.</span><span class=\"T_dna\">.</span><span class=\"A_dna\">.</span><span class=\"C_dna\">.</span><span class=\"T_dna\">.</span><span class=\"C_dna\">.</span><span class=\"A_dna\">.</span><span class=\"T_dna\">.</span><span class=\"G_dna\">.</span><span class=\"C_dna\">.</span><span class=\"C_dna\">.</span><span class=\"A_dna\">.</span><span class=\"G_dna\">.</span><span class=\"C_dna\">.</span><span class=\"T_dna\">.</span><span class=\"C_dna\">.</span><span class=\"A_dna\">.</span><span class=\"T_dna\">.</span><span class=\"T_dna\">.</span><span class=\"A_dna\">.</span><span class=\"C_dna\">.</span><span class=\"A_dna\">.</span><span class=\"G_dna\">.</span><span class=\"C_dna\">.</span><span class=\"A_dna\">.</span><span class=\"T_dna\">.</span><span class=\"G_dna\">.</span><span class=\"A_dna\">.</span><span class=\"G_dna\">.</span><span class=\"A_dna\">.</span><span class=\"A_dna\">.</span><span class=\"C_dna\">.</span><span class=\"A_dna\">.</span><span class=\"G_dna\">.</span><span class=\"C_dna\">.</span><span class=\"A_dna\">.</span><span class=\"G_dna\">.</span><span class=\"T_dna\">.</span><span class=\"T_dna\">.</span><span class=\"T_dna\">.</span><span class=\"A_dna\">.</span><span class=\"T_dna\">.</span><span class=\"T_dna\">.</span><span class=\"A_dna\">.</span><span class=\"C_dna\">.</span><span class=\"T_dna\">.</span><span class=\"C_dna\">.</span><span class=\"A_dna\">.</span><span class=\"C_dna\">.</span><span class=\"T_dna\">.</span></td></tr>\n",
       "<tr><td class=\"label\">Gorilla</td><td><span class=\"T_dna\">.</span><span class=\"G_dna\">.</span><span class=\"T_dna\">.</span><span class=\"G_dna\">.</span><span class=\"G_dna\">.</span><span class=\"C_dna\">.</span><span class=\"A_dna\">.</span><span class=\"C_dna\">.</span><span class=\"A_dna\">.</span><span class=\"A_dna\">.</span><span class=\"A_dna\">.</span><span class=\"T_dna\">.</span><span class=\"A_dna\">.</span><span class=\"C_dna\">.</span><span class=\"T_dna\">.</span><span class=\"C_dna\">.</span><span class=\"A_dna\">.</span><span class=\"T_dna\">.</span><span class=\"G_dna\">.</span><span class=\"C_dna\">.</span><span class=\"C_dna\">.</span><span class=\"A_dna\">.</span><span class=\"G_dna\">.</span><span class=\"C_dna\">.</span><span class=\"T_dna\">.</span><span class=\"C_dna\">.</span><span class=\"A_dna\">.</span><span class=\"T_dna\">.</span><span class=\"T_dna\">.</span><span class=\"A_dna\">.</span><span class=\"C_dna\">.</span><span class=\"A_dna\">.</span><span class=\"G_dna\">.</span><span class=\"C_dna\">.</span><span class=\"A_dna\">.</span><span class=\"T_dna\">.</span><span class=\"G_dna\">.</span><span class=\"A_dna\">.</span><span class=\"G_dna\">.</span><span class=\"A_dna\">.</span><span class=\"A_dna\">.</span><span class=\"C_dna\">.</span><span class=\"A_dna\">.</span><span class=\"G_dna\">.</span><span class=\"C_dna\">.</span><span class=\"A_dna\">.</span><span class=\"G_dna\">.</span><span class=\"T_dna\">.</span><span class=\"T_dna\">.</span><span class=\"T_dna\">.</span><span class=\"A_dna\">.</span><span class=\"T_dna\">.</span><span class=\"T_dna\">.</span><span class=\"A_dna\">.</span><span class=\"C_dna\">.</span><span class=\"T_dna\">.</span><span class=\"C_dna\">.</span><span class=\"A_dna\">.</span><span class=\"C_dna\">.</span><span class=\"T_dna\">.</span></td></tr>\n",
       "</table>\n",
       "<p><i>3 x 2838 (truncated to 3 x 60) dna alignment</i></p>\n",
       "</div>"
      ],
      "text/plain": [
       "3 x 2838 dna alignment: Chimpanzee[TGTGGCACAA...], Human[TGTGGCACAA...], Gorilla[TGTGGCACAA...]"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "aln = aln.take_seqs([\"Chimpanzee\", \"Human\", \"Gorilla\"])\n",
    "aln1 = aln.omit_gap_pos(motif_length=3, allowed_gap_frac=0.0)\n",
    "aln1"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5edca7bf",
   "metadata": {},
   "source": [
    "## Context dependent model's have a frame\n",
    "\n",
    "This is obviously True for codon models in so much as an alignment is split into non-overlapping trinucleotides and those are modelled. But it's also true for dinucleotide in that the alignment is split into non-verlapping dinucleotides. We can check the robustness of the fit to *frame* by evaluating the other frame (slicing from 1). I should just point out that there is no known biological basis for the concept of a dinucleotide frame. It is purely something to consider with respect to data sampling and parameter estimation."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "4aaace54",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<style>\n",
       ".c3align table {margin: 10px 0;}\n",
       ".c3align td { border: none !important; text-align: left !important; }\n",
       ".c3align tr:not(.num_row) td span {margin: 0 2px;}\n",
       ".c3align tr:nth-child(even) {background: #f7f7f7;}\n",
       ".c3align .num_row {background-color:rgba(161, 195, 209, 0.5) !important; border-top: solid 1px black; }\n",
       ".c3align .label { font-size: 12pt ; text-align: right !important; color: black !important; padding: 0 4px; display: table-cell !important; font-weight: normal !important; }\n",
       ".c3align .T_dna{font-family: \"Lucida Console\",monospace !important; font-size: 12pt !important; color: blue; }\n",
       ".c3align .C_dna{font-family: \"Lucida Console\",monospace !important; font-size: 12pt !important; color: black; }\n",
       ".c3align .A_dna{font-family: \"Lucida Console\",monospace !important; font-size: 12pt !important; color: #FF0102; }\n",
       ".c3align .G_dna{font-family: \"Lucida Console\",monospace !important; font-size: 12pt !important; color: green; }\n",
       ".c3align .terminal_ambig_dna{font-family: \"Lucida Console\",monospace !important; font-size: 12pt !important; color: gray; }\n",
       ".c3align .ambig_dna{font-family: \"Lucida Console\",monospace !important; font-size: 12pt !important; color: gray; }\n",
       "</style>\n",
       "<div class=\"c3align\">\n",
       "<table>\n",
       "<tr class=\"num_row\"><td></td><td><b>0</b></td></tr>\n",
       "<tr><td class=\"label\">Chimpanzee</td><td><span class=\"G_dna\">G</span><span class=\"T_dna\">T</span><span class=\"G_dna\">G</span><span class=\"G_dna\">G</span><span class=\"C_dna\">C</span><span class=\"A_dna\">A</span><span class=\"C_dna\">C</span><span class=\"A_dna\">A</span><span class=\"A_dna\">A</span><span class=\"A_dna\">A</span><span class=\"T_dna\">T</span><span class=\"A_dna\">A</span><span class=\"C_dna\">C</span><span class=\"T_dna\">T</span><span class=\"C_dna\">C</span><span class=\"A_dna\">A</span><span class=\"T_dna\">T</span><span class=\"G_dna\">G</span><span class=\"C_dna\">C</span><span class=\"C_dna\">C</span><span class=\"A_dna\">A</span><span class=\"G_dna\">G</span><span class=\"C_dna\">C</span><span class=\"T_dna\">T</span><span class=\"C_dna\">C</span><span class=\"A_dna\">A</span><span class=\"T_dna\">T</span><span class=\"T_dna\">T</span><span class=\"A_dna\">A</span><span class=\"C_dna\">C</span><span class=\"A_dna\">A</span><span class=\"G_dna\">G</span><span class=\"C_dna\">C</span><span class=\"A_dna\">A</span><span class=\"T_dna\">T</span><span class=\"G_dna\">G</span><span class=\"A_dna\">A</span><span class=\"G_dna\">G</span><span class=\"A_dna\">A</span><span class=\"A_dna\">A</span><span class=\"C_dna\">C</span><span class=\"A_dna\">A</span><span class=\"G_dna\">G</span><span class=\"C_dna\">C</span><span class=\"A_dna\">A</span><span class=\"G_dna\">G</span><span class=\"T_dna\">T</span><span class=\"T_dna\">T</span><span class=\"T_dna\">T</span><span class=\"A_dna\">A</span><span class=\"T_dna\">T</span><span class=\"T_dna\">T</span><span class=\"A_dna\">A</span><span class=\"C_dna\">C</span><span class=\"T_dna\">T</span><span class=\"C_dna\">C</span><span class=\"A_dna\">A</span><span class=\"C_dna\">C</span><span class=\"T_dna\">T</span><span class=\"A_dna\">A</span></td></tr>\n",
       "<tr><td class=\"label\">Human</td><td><span class=\"G_dna\">.</span><span class=\"T_dna\">.</span><span class=\"G_dna\">.</span><span class=\"G_dna\">.</span><span class=\"C_dna\">.</span><span class=\"A_dna\">.</span><span class=\"C_dna\">.</span><span class=\"A_dna\">.</span><span class=\"A_dna\">.</span><span class=\"A_dna\">.</span><span class=\"T_dna\">.</span><span class=\"A_dna\">.</span><span class=\"C_dna\">.</span><span class=\"T_dna\">.</span><span class=\"C_dna\">.</span><span class=\"A_dna\">.</span><span class=\"T_dna\">.</span><span class=\"G_dna\">.</span><span class=\"C_dna\">.</span><span class=\"C_dna\">.</span><span class=\"A_dna\">.</span><span class=\"G_dna\">.</span><span class=\"C_dna\">.</span><span class=\"T_dna\">.</span><span class=\"C_dna\">.</span><span class=\"A_dna\">.</span><span class=\"T_dna\">.</span><span class=\"T_dna\">.</span><span class=\"A_dna\">.</span><span class=\"C_dna\">.</span><span class=\"A_dna\">.</span><span class=\"G_dna\">.</span><span class=\"C_dna\">.</span><span class=\"A_dna\">.</span><span class=\"T_dna\">.</span><span class=\"G_dna\">.</span><span class=\"A_dna\">.</span><span class=\"G_dna\">.</span><span class=\"A_dna\">.</span><span class=\"A_dna\">.</span><span class=\"C_dna\">.</span><span class=\"A_dna\">.</span><span class=\"G_dna\">.</span><span class=\"C_dna\">.</span><span class=\"A_dna\">.</span><span class=\"G_dna\">.</span><span class=\"T_dna\">.</span><span class=\"T_dna\">.</span><span class=\"T_dna\">.</span><span class=\"A_dna\">.</span><span class=\"T_dna\">.</span><span class=\"T_dna\">.</span><span class=\"A_dna\">.</span><span class=\"C_dna\">.</span><span class=\"T_dna\">.</span><span class=\"C_dna\">.</span><span class=\"A_dna\">.</span><span class=\"C_dna\">.</span><span class=\"T_dna\">.</span><span class=\"A_dna\">.</span></td></tr>\n",
       "<tr><td class=\"label\">Gorilla</td><td><span class=\"G_dna\">.</span><span class=\"T_dna\">.</span><span class=\"G_dna\">.</span><span class=\"G_dna\">.</span><span class=\"C_dna\">.</span><span class=\"A_dna\">.</span><span class=\"C_dna\">.</span><span class=\"A_dna\">.</span><span class=\"A_dna\">.</span><span class=\"A_dna\">.</span><span class=\"T_dna\">.</span><span class=\"A_dna\">.</span><span class=\"C_dna\">.</span><span class=\"T_dna\">.</span><span class=\"C_dna\">.</span><span class=\"A_dna\">.</span><span class=\"T_dna\">.</span><span class=\"G_dna\">.</span><span class=\"C_dna\">.</span><span class=\"C_dna\">.</span><span class=\"A_dna\">.</span><span class=\"G_dna\">.</span><span class=\"C_dna\">.</span><span class=\"T_dna\">.</span><span class=\"C_dna\">.</span><span class=\"A_dna\">.</span><span class=\"T_dna\">.</span><span class=\"T_dna\">.</span><span class=\"A_dna\">.</span><span class=\"C_dna\">.</span><span class=\"A_dna\">.</span><span class=\"G_dna\">.</span><span class=\"C_dna\">.</span><span class=\"A_dna\">.</span><span class=\"T_dna\">.</span><span class=\"G_dna\">.</span><span class=\"A_dna\">.</span><span class=\"G_dna\">.</span><span class=\"A_dna\">.</span><span class=\"A_dna\">.</span><span class=\"C_dna\">.</span><span class=\"A_dna\">.</span><span class=\"G_dna\">.</span><span class=\"C_dna\">.</span><span class=\"A_dna\">.</span><span class=\"G_dna\">.</span><span class=\"T_dna\">.</span><span class=\"T_dna\">.</span><span class=\"T_dna\">.</span><span class=\"A_dna\">.</span><span class=\"T_dna\">.</span><span class=\"T_dna\">.</span><span class=\"A_dna\">.</span><span class=\"C_dna\">.</span><span class=\"T_dna\">.</span><span class=\"C_dna\">.</span><span class=\"A_dna\">.</span><span class=\"C_dna\">.</span><span class=\"T_dna\">.</span><span class=\"A_dna\">.</span></td></tr>\n",
       "</table>\n",
       "<p><i>3 x 2828 (truncated to 3 x 60) dna alignment</i></p>\n",
       "</div>"
      ],
      "text/plain": [
       "3 x 2828 dna alignment: Chimpanzee[GTGGCACAAA...], Human[GTGGCACAAA...], Gorilla[GTGGCACAAA...]"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "aln2 = aln[1:].omit_gap_pos(motif_length=2, allowed_gap_frac=0.0)\n",
    "aln2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "acfe0ebc",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "model(sm=NonReversibleDinucleotide(name='GDN_CpG_ss'; params=['A>C', 'A>T',\n",
       "'A>G', 'C>A', 'C>T', 'C>G', 'T>A', 'T>C', 'G>A', 'G>C', 'G>T', '(CG>TG |\n",
       "CG>CA)']; num_motifs=16; motifs=['TT', 'TC', 'TA', 'TG', 'CT', 'CC', 'CA', 'CG',\n",
       "'AT', 'AC', 'AA', 'AG', 'GT', 'GC', 'GA', 'GG'])), tree=None,\n",
       "unique_trees=False, tree_func=None, name=None, optimise_motif_probs=True,\n",
       "sm_args=None, lf_args=None, time_het='max', param_rules=None, opt_args=None,\n",
       "lower=1e-06, upper=50, split_codons=False, show_progress=True, verbose=False)"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "di_cpg = c3.get_app(\"model\", GDN_CpG_ss(), time_het=\"max\", show_progress=True, optimise_motif_probs=True)\n",
    "di_cpg"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "44064f61",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "fc7d713feca84a85b586f1301d24fe25",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "   0%|          |00:00<?"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "fc6d5a8ade4b4881a0a603221b0126ac",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "   0%|          |00:00<?"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<h4>GDN_CpG_ss</h4>\n",
       "<p>log-likelihood = -3993.1753</p>\n",
       "<p>number of free parameters = 54</p>\n",
       "\n",
       "<div class=\"c3table\">\n",
       "\n",
       "<style>\n",
       ".c3table table {margin: 10px 0;}\n",
       ".c3table tr:last-child {border-bottom: 1px solid #000;} \n",
       ".c3table tr > th {text-align: left; padding: 0 5px;}\n",
       ".c3table tr > td {text-align: left; padding: 5px;}\n",
       ".c3table tr:nth-child(even) {background: #f7f7f7 !important;}\n",
       ".c3table .ellipsis {background: rgba(0, 0, 0, .01);}\n",
       ".c3table .index {background: rgba(161, 195, 209, 0.0); margin: 10px; font-weight: 600;}\n",
       ".c3table .head_cell {background: rgba(161, 195, 209, 0.75); font-weight: bold; text-align: center;}\n",
       ".c3table caption {color: rgb(250, 250, 250); background: rgba(30, 140, 200, 1); padding: 3px; white-space: nowrap; caption-side: top;}\n",
       ".c3table .cell_title {font-weight: bold;}\n",
       ".c3col_left { text-align: left !important; display: block;}\n",
       ".c3col_right { text-align: right !important; display: block;}\n",
       ".c3col_center { text-align: center !important; display: block;}\n",
       "</style>\n",
       "\n",
       "<table>\n",
       "\n",
       "<caption>\n",
       "<span class=\"cell_title\">Edge params</span>\n",
       "</caption>\n",
       "<thead class=\"head_cell\">\n",
       "<th>edge</th><th>parent</th><th>length</th><th>(CG>TG | CG>CA)</th><th>A>C</th><th>A>G</th><th>A>T</th><th>C>A</th><th>C>G</th>\n",
       "</thead>\n",
       "<tbody>\n",
       "<tr><td><span class=\"c3col_left\">Chimpanzee</span></td><td><span class=\"c3col_left\">root</span></td><td><span class=\"c3col_right\">0.01</span></td><td><span class=\"c3col_right\">0.00</span></td><td><span class=\"c3col_right\">0.00</span></td><td><span class=\"c3col_right\">10.81</span></td><td><span class=\"c3col_right\">0.00</span></td><td><span class=\"c3col_right\">0.00</span></td><td><span class=\"c3col_right\">23.38</span></td></tr>\n",
       "<tr><td><span class=\"c3col_left\">Human</span></td><td><span class=\"c3col_left\">root</span></td><td><span class=\"c3col_right\">0.01</span></td><td><span class=\"c3col_right\">7.44</span></td><td><span class=\"c3col_right\">0.00</span></td><td><span class=\"c3col_right\">1.96</span></td><td><span class=\"c3col_right\">0.00</span></td><td><span class=\"c3col_right\">1.38</span></td><td><span class=\"c3col_right\">2.78</span></td></tr>\n",
       "<tr><td><span class=\"c3col_left\">Gorilla</span></td><td><span class=\"c3col_left\">root</span></td><td><span class=\"c3col_right\">0.00</span></td><td><span class=\"c3col_right\">0.00</span></td><td><span class=\"c3col_right\">14.14</span></td><td><span class=\"c3col_right\">42.55</span></td><td><span class=\"c3col_right\">0.00</span></td><td><span class=\"c3col_right\">30.18</span></td><td><span class=\"c3col_right\">0.00</span></td></tr>\n",
       "</tbody>\n",
       "</table>\n",
       "\n",
       "<table>\n",
       "\n",
       "<caption>\n",
       "<span class=\"cell_title\">continuation</span>\n",
       "</caption>\n",
       "<thead class=\"head_cell\">\n",
       "<th>C>T</th><th>G>A</th><th>G>C</th><th>G>T</th><th>T>A</th><th>T>C</th>\n",
       "</thead>\n",
       "<tbody>\n",
       "<tr><td><span class=\"c3col_right\">24.06</span></td><td><span class=\"c3col_right\">19.64</span></td><td><span class=\"c3col_right\">19.23</span></td><td><span class=\"c3col_right\">0.00</span></td><td><span class=\"c3col_right\">0.00</span></td><td><span class=\"c3col_right\">50.00</span></td></tr>\n",
       "<tr><td><span class=\"c3col_right\">2.39</span></td><td><span class=\"c3col_right\">5.05</span></td><td><span class=\"c3col_right\">1.14</span></td><td><span class=\"c3col_right\">0.00</span></td><td><span class=\"c3col_right\">1.00</span></td><td><span class=\"c3col_right\">2.02</span></td></tr>\n",
       "<tr><td><span class=\"c3col_right\">0.00</span></td><td><span class=\"c3col_right\">50.00</span></td><td><span class=\"c3col_right\">0.00</span></td><td><span class=\"c3col_right\">0.00</span></td><td><span class=\"c3col_right\">0.00</span></td><td><span class=\"c3col_right\">0.00</span></td></tr>\n",
       "</tbody>\n",
       "</table>\n",
       "\n",
       "</div>\n",
       "\n",
       "<div class=\"c3table\">\n",
       "\n",
       "<style>\n",
       ".c3table table {margin: 10px 0;}\n",
       ".c3table tr:last-child {border-bottom: 1px solid #000;} \n",
       ".c3table tr > th {text-align: left; padding: 0 5px;}\n",
       ".c3table tr > td {text-align: left; padding: 5px;}\n",
       ".c3table tr:nth-child(even) {background: #f7f7f7 !important;}\n",
       ".c3table .ellipsis {background: rgba(0, 0, 0, .01);}\n",
       ".c3table .index {background: rgba(161, 195, 209, 0.0); margin: 10px; font-weight: 600;}\n",
       ".c3table .head_cell {background: rgba(161, 195, 209, 0.75); font-weight: bold; text-align: center;}\n",
       ".c3table caption {color: rgb(250, 250, 250); background: rgba(30, 140, 200, 1); padding: 3px; white-space: nowrap; caption-side: top;}\n",
       ".c3table .cell_title {font-weight: bold;}\n",
       ".c3col_left { text-align: left !important; display: block;}\n",
       ".c3col_right { text-align: right !important; display: block;}\n",
       ".c3col_center { text-align: center !important; display: block;}\n",
       "</style>\n",
       "\n",
       "<table>\n",
       "\n",
       "<caption>\n",
       "<span class=\"cell_title\">Motif params</span>\n",
       "</caption>\n",
       "<thead class=\"head_cell\">\n",
       "<th>AA</th><th>AC</th><th>AG</th><th>AT</th><th>CA</th><th>CC</th><th>CG</th><th>CT</th><th>GA</th><th>GC</th><th>GG</th><th>GT</th><th>TA</th>\n",
       "</thead>\n",
       "<tbody>\n",
       "<tr><td><span class=\"c3col_right\">0.14</span></td><td><span class=\"c3col_right\">0.05</span></td><td><span class=\"c3col_right\">0.11</span></td><td><span class=\"c3col_right\">0.08</span></td><td><span class=\"c3col_right\">0.07</span></td><td><span class=\"c3col_right\">0.05</span></td><td><span class=\"c3col_right\">0.01</span></td><td><span class=\"c3col_right\">0.05</span></td><td><span class=\"c3col_right\">0.08</span></td><td><span class=\"c3col_right\">0.04</span></td><td><span class=\"c3col_right\">0.04</span></td><td><span class=\"c3col_right\">0.05</span></td><td><span class=\"c3col_right\">0.07</span></td></tr>\n",
       "</tbody>\n",
       "</table>\n",
       "\n",
       "<table>\n",
       "\n",
       "<caption>\n",
       "<span class=\"cell_title\">continuation</span>\n",
       "</caption>\n",
       "<thead class=\"head_cell\">\n",
       "<th>TC</th><th>TG</th><th>TT</th>\n",
       "</thead>\n",
       "<tbody>\n",
       "<tr><td><span class=\"c3col_right\">0.04</span></td><td><span class=\"c3col_right\">0.06</span></td><td><span class=\"c3col_right\">0.06</span></td></tr>\n",
       "</tbody>\n",
       "</table>\n",
       "\n",
       "</div>"
      ],
      "text/plain": [
       "GDN_CpG_ss\n",
       "log-likelihood = -3993.1753\n",
       "number of free parameters = 54\n",
       "===========================================================================\n",
       "edge          parent    length    (CG>TG | CG>CA)      A>C      A>G     A>T\n",
       "---------------------------------------------------------------------------\n",
       "Chimpanzee    root        0.01               0.00     0.00    10.81    0.00\n",
       "Human         root        0.01               7.44     0.00     1.96    0.00\n",
       "Gorilla       root        0.00               0.00    14.14    42.55    0.00\n",
       "---------------------------------------------------------------------------\n",
       "\n",
       "continued: \n",
       "==================================================================\n",
       "  C>A      C>G      C>T      G>A      G>C     G>T     T>A      T>C\n",
       "------------------------------------------------------------------\n",
       " 0.00    23.38    24.06    19.64    19.23    0.00    0.00    50.00\n",
       " 1.38     2.78     2.39     5.05     1.14    0.00    1.00     2.02\n",
       "30.18     0.00     0.00    50.00     0.00    0.00    0.00     0.00\n",
       "------------------------------------------------------------------\n",
       "\n",
       "============================================================================\n",
       "  AA      AC      AG      AT      CA      CC      CG      CT      GA      GC\n",
       "----------------------------------------------------------------------------\n",
       "0.14    0.05    0.11    0.08    0.07    0.05    0.01    0.05    0.08    0.04\n",
       "----------------------------------------------------------------------------\n",
       "\n",
       "continued: \n",
       "============================================\n",
       "  GG      GT      TA      TC      TG      TT\n",
       "--------------------------------------------\n",
       "0.04    0.05    0.07    0.04    0.06    0.06\n",
       "--------------------------------------------"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "result1 = di_cpg(aln1)\n",
    "result1.lf"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "1bc3f1a0",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "6ab1a1aa0c0748e799ea6dda8ff450fe",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "   0%|          |00:00<?"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "34efd4798569453fb08b38ca68d1b83d",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "   0%|          |00:00<?"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<h4>GDN_CpG_ss</h4>\n",
       "<p>log-likelihood = -3971.4674</p>\n",
       "<p>number of free parameters = 54</p>\n",
       "\n",
       "<div class=\"c3table\">\n",
       "\n",
       "<style>\n",
       ".c3table table {margin: 10px 0;}\n",
       ".c3table tr:last-child {border-bottom: 1px solid #000;} \n",
       ".c3table tr > th {text-align: left; padding: 0 5px;}\n",
       ".c3table tr > td {text-align: left; padding: 5px;}\n",
       ".c3table tr:nth-child(even) {background: #f7f7f7 !important;}\n",
       ".c3table .ellipsis {background: rgba(0, 0, 0, .01);}\n",
       ".c3table .index {background: rgba(161, 195, 209, 0.0); margin: 10px; font-weight: 600;}\n",
       ".c3table .head_cell {background: rgba(161, 195, 209, 0.75); font-weight: bold; text-align: center;}\n",
       ".c3table caption {color: rgb(250, 250, 250); background: rgba(30, 140, 200, 1); padding: 3px; white-space: nowrap; caption-side: top;}\n",
       ".c3table .cell_title {font-weight: bold;}\n",
       ".c3col_left { text-align: left !important; display: block;}\n",
       ".c3col_right { text-align: right !important; display: block;}\n",
       ".c3col_center { text-align: center !important; display: block;}\n",
       "</style>\n",
       "\n",
       "<table>\n",
       "\n",
       "<caption>\n",
       "<span class=\"cell_title\">Edge params</span>\n",
       "</caption>\n",
       "<thead class=\"head_cell\">\n",
       "<th>edge</th><th>parent</th><th>length</th><th>(CG>TG | CG>CA)</th><th>A>C</th><th>A>G</th><th>A>T</th><th>C>A</th><th>C>G</th>\n",
       "</thead>\n",
       "<tbody>\n",
       "<tr><td><span class=\"c3col_left\">Chimpanzee</span></td><td><span class=\"c3col_left\">root</span></td><td><span class=\"c3col_right\">0.01</span></td><td><span class=\"c3col_right\">0.00</span></td><td><span class=\"c3col_right\">0.00</span></td><td><span class=\"c3col_right\">10.87</span></td><td><span class=\"c3col_right\">0.00</span></td><td><span class=\"c3col_right\">0.00</span></td><td><span class=\"c3col_right\">23.49</span></td></tr>\n",
       "<tr><td><span class=\"c3col_left\">Human</span></td><td><span class=\"c3col_left\">root</span></td><td><span class=\"c3col_right\">0.01</span></td><td><span class=\"c3col_right\">10.99</span></td><td><span class=\"c3col_right\">0.00</span></td><td><span class=\"c3col_right\">1.96</span></td><td><span class=\"c3col_right\">0.00</span></td><td><span class=\"c3col_right\">1.38</span></td><td><span class=\"c3col_right\">2.79</span></td></tr>\n",
       "<tr><td><span class=\"c3col_left\">Gorilla</span></td><td><span class=\"c3col_left\">root</span></td><td><span class=\"c3col_right\">0.00</span></td><td><span class=\"c3col_right\">0.00</span></td><td><span class=\"c3col_right\">14.14</span></td><td><span class=\"c3col_right\">42.44</span></td><td><span class=\"c3col_right\">0.00</span></td><td><span class=\"c3col_right\">30.32</span></td><td><span class=\"c3col_right\">0.00</span></td></tr>\n",
       "</tbody>\n",
       "</table>\n",
       "\n",
       "<table>\n",
       "\n",
       "<caption>\n",
       "<span class=\"cell_title\">continuation</span>\n",
       "</caption>\n",
       "<thead class=\"head_cell\">\n",
       "<th>C>T</th><th>G>A</th><th>G>C</th><th>G>T</th><th>T>A</th><th>T>C</th>\n",
       "</thead>\n",
       "<tbody>\n",
       "<tr><td><span class=\"c3col_right\">24.05</span></td><td><span class=\"c3col_right\">19.83</span></td><td><span class=\"c3col_right\">19.36</span></td><td><span class=\"c3col_right\">0.00</span></td><td><span class=\"c3col_right\">0.00</span></td><td><span class=\"c3col_right\">50.00</span></td></tr>\n",
       "<tr><td><span class=\"c3col_right\">2.38</span></td><td><span class=\"c3col_right\">5.05</span></td><td><span class=\"c3col_right\">1.14</span></td><td><span class=\"c3col_right\">0.00</span></td><td><span class=\"c3col_right\">1.00</span></td><td><span class=\"c3col_right\">2.01</span></td></tr>\n",
       "<tr><td><span class=\"c3col_right\">0.00</span></td><td><span class=\"c3col_right\">50.00</span></td><td><span class=\"c3col_right\">0.00</span></td><td><span class=\"c3col_right\">0.00</span></td><td><span class=\"c3col_right\">0.00</span></td><td><span class=\"c3col_right\">0.00</span></td></tr>\n",
       "</tbody>\n",
       "</table>\n",
       "\n",
       "</div>\n",
       "\n",
       "<div class=\"c3table\">\n",
       "\n",
       "<style>\n",
       ".c3table table {margin: 10px 0;}\n",
       ".c3table tr:last-child {border-bottom: 1px solid #000;} \n",
       ".c3table tr > th {text-align: left; padding: 0 5px;}\n",
       ".c3table tr > td {text-align: left; padding: 5px;}\n",
       ".c3table tr:nth-child(even) {background: #f7f7f7 !important;}\n",
       ".c3table .ellipsis {background: rgba(0, 0, 0, .01);}\n",
       ".c3table .index {background: rgba(161, 195, 209, 0.0); margin: 10px; font-weight: 600;}\n",
       ".c3table .head_cell {background: rgba(161, 195, 209, 0.75); font-weight: bold; text-align: center;}\n",
       ".c3table caption {color: rgb(250, 250, 250); background: rgba(30, 140, 200, 1); padding: 3px; white-space: nowrap; caption-side: top;}\n",
       ".c3table .cell_title {font-weight: bold;}\n",
       ".c3col_left { text-align: left !important; display: block;}\n",
       ".c3col_right { text-align: right !important; display: block;}\n",
       ".c3col_center { text-align: center !important; display: block;}\n",
       "</style>\n",
       "\n",
       "<table>\n",
       "\n",
       "<caption>\n",
       "<span class=\"cell_title\">Motif params</span>\n",
       "</caption>\n",
       "<thead class=\"head_cell\">\n",
       "<th>AA</th><th>AC</th><th>AG</th><th>AT</th><th>CA</th><th>CC</th><th>CG</th><th>CT</th><th>GA</th><th>GC</th><th>GG</th><th>GT</th><th>TA</th>\n",
       "</thead>\n",
       "<tbody>\n",
       "<tr><td><span class=\"c3col_right\">0.15</span></td><td><span class=\"c3col_right\">0.05</span></td><td><span class=\"c3col_right\">0.10</span></td><td><span class=\"c3col_right\">0.07</span></td><td><span class=\"c3col_right\">0.08</span></td><td><span class=\"c3col_right\">0.03</span></td><td><span class=\"c3col_right\">0.01</span></td><td><span class=\"c3col_right\">0.06</span></td><td><span class=\"c3col_right\">0.08</span></td><td><span class=\"c3col_right\">0.04</span></td><td><span class=\"c3col_right\">0.05</span></td><td><span class=\"c3col_right\">0.05</span></td><td><span class=\"c3col_right\">0.07</span></td></tr>\n",
       "</tbody>\n",
       "</table>\n",
       "\n",
       "<table>\n",
       "\n",
       "<caption>\n",
       "<span class=\"cell_title\">continuation</span>\n",
       "</caption>\n",
       "<thead class=\"head_cell\">\n",
       "<th>TC</th><th>TG</th><th>TT</th>\n",
       "</thead>\n",
       "<tbody>\n",
       "<tr><td><span class=\"c3col_right\">0.05</span></td><td><span class=\"c3col_right\">0.06</span></td><td><span class=\"c3col_right\">0.06</span></td></tr>\n",
       "</tbody>\n",
       "</table>\n",
       "\n",
       "</div>"
      ],
      "text/plain": [
       "GDN_CpG_ss\n",
       "log-likelihood = -3971.4674\n",
       "number of free parameters = 54\n",
       "===========================================================================\n",
       "edge          parent    length    (CG>TG | CG>CA)      A>C      A>G     A>T\n",
       "---------------------------------------------------------------------------\n",
       "Chimpanzee    root        0.01               0.00     0.00    10.87    0.00\n",
       "Human         root        0.01              10.99     0.00     1.96    0.00\n",
       "Gorilla       root        0.00               0.00    14.14    42.44    0.00\n",
       "---------------------------------------------------------------------------\n",
       "\n",
       "continued: \n",
       "==================================================================\n",
       "  C>A      C>G      C>T      G>A      G>C     G>T     T>A      T>C\n",
       "------------------------------------------------------------------\n",
       " 0.00    23.49    24.05    19.83    19.36    0.00    0.00    50.00\n",
       " 1.38     2.79     2.38     5.05     1.14    0.00    1.00     2.01\n",
       "30.32     0.00     0.00    50.00     0.00    0.00    0.00     0.00\n",
       "------------------------------------------------------------------\n",
       "\n",
       "============================================================================\n",
       "  AA      AC      AG      AT      CA      CC      CG      CT      GA      GC\n",
       "----------------------------------------------------------------------------\n",
       "0.15    0.05    0.10    0.07    0.08    0.03    0.01    0.06    0.08    0.04\n",
       "----------------------------------------------------------------------------\n",
       "\n",
       "continued: \n",
       "============================================\n",
       "  GG      GT      TA      TC      TG      TT\n",
       "--------------------------------------------\n",
       "0.05    0.05    0.07    0.05    0.06    0.06\n",
       "--------------------------------------------"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "result2 = di_cpg(aln2)\n",
    "result2.lf"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "215feef2",
   "metadata": {},
   "source": [
    "TO DO: a codons variant of the above, change the base model class and add \"omega\" (see my earlier papers on this)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "c313",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.13.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
